### ISSUE TITLE
An error is thrown when maximizing the Scene/Game view

### ISSUE THREAD

**Issue by epicsquad** [handzlikchris/FastScriptReload#141]
After installing FastScriptReload using package manager. I get error when trying to maximize Game view I get console error:
Invalid editor window of type: FastScriptReload.Editor.FastScriptReloadWelcomeScreen, title: FastScriptReload.Editor.FastScriptReloadWelcomeScreen
UnityEditor.EditorApplication:Internal_CallDelayFunctions () (at /Users/bokken/build/output/unity/unity/Editor/Mono/EditorApplication.cs:377)

I googled that type of error and all recommends to reset Layout to factory settings or save your current layout and reactivate it every time you get this error and this works find, but when you enter or exit play mode each time you should reactive your layout or you got this error. May be someone know works around for this issue? Or this is on of typical classical Unity bugs that knows, but was never fixed.

**Comment by Infinite-3D** [comment]
> Doesn't seem to be causing any issues though - i think we'll have to just leave it as is.

To someone who uses the "Error Pause" setting in Unity, this is quite annoying as this error automatically pauses the game and exits full screen whenever I enter it. Any way this can be thrown in a try catch block or something?

**Comment by RunninglVlan** [comment]
This also breaks some of our Play Mode tests, as we're doing Screenshots of Fullscreen Editor windows. Although now that I think about, we could try temporarily ignoring errors with `LogAssert.ignoreFailingMessages`

**Comment by OddHero** [comment]
So it annoyed me enough to dig into this.....

The problem is triggered by UnityEditor.WindowLayout
```csharp
internal static void CheckWindowConsistency()
{
  foreach (EditorWindow editorWindow in Resources.FindObjectsOfTypeAll(typeof (EditorWindow)))
  {
    if ((UnityEngine.Object) editorWindow.m_Parent == (UnityEngine.Object) null)
      Debug.LogErrorFormat("Invalid editor window of type: {0}, title: {1}", (object) editorWindow.GetType(), (object) editorWindow.titleContent.text);
  }
}
```

So the actual problem is that something keeps an instance of `Editor.FastScriptReloadWelcomeScreen` alive that is not registered in the current WindowLayout... So basically somewhere there has to be an orphaned EditorWindow-Instance...

So I dug deeper and found that the following Method which runs from the static constructor of the window
`InitCommon()->AutoDetectAndSetShaderMode()->SetCommonMaterialsShader(usedShaderMode);`
```csharp
public static void SetCommonMaterialsShader(ShadersMode newShaderModeValue)
        {
            var rootToolFolder = AssetPathResolver.GetAssetFolderPathRelativeToScript(ScriptableObject.CreateInstance(typeof(FastScriptReloadWelcomeScreen)), 1);
            if (rootToolFolder.Contains("/Scripts"))
            {
                rootToolFolder = rootToolFolder.Replace("/Scripts", ""); //if nested remove that and go higher level
            }
            var assets = AssetDatabase.FindAssets("t:Material Point", new[] { rootToolFolder });
...
```
uses `ScriptableObject.CreateInstance` to create an instance of the EditorWindow... Sus... 😅

So part one of my Hotfix:
```csharp
public static void SetCommonMaterialsShader(ShadersMode newShaderModeValue)
{
    return;
}

```

Unfortunately that is not the only problem....
The second part of the Problem lies in the Way the editor window is created...
I do not fully why the following replacement for the Init() method fixes the Problem, but Ill try to explain what I think is happening
```csharp
        public static FastScriptReloadWelcomeScreen Init()
        {
            EditorWindow window = GetWindow(typeof(FastScriptReloadWelcomeScreen),true, _WindowTitle, true);
            window.minSize = _WindowSizePx;
            window.maxSize = _WindowSizePx;
            return window as FastScriptReloadWelcomeScreen;
        }
```

The reason why I replaced the Method in the first place was that while digging thru the trace, I found this one:
```csharp
    public static object OpenWindow(System.Type windowType, string title, Vector2 windowSizePx)
    {
      EditorWindow window = EditorWindow.GetWindow(windowType, true, title);
      window.minSize = windowSizePx;
      window.maxSize = windowSizePx;
      window.Show();
      ProductWelcomeScreenBase._ContentPanelWidthPx = (int) windowSizePx.x - 145;
      return (object) window;
    }
```

I thought maybe the problem is that the window is created with the utility flag set but then .Show() is used to display the window. There is ShowUtility for that so i thought lets try to replace this with "normal" EditorWindow syntax from the docs:
```csharp
[MenuItem("Examples/My Editor Window")]
public static void ShowExample()
{
    MyEditorWindow wnd = GetWindow<MyEditorWindow>();
    wnd.titleContent = new GUIContent("MyEditorWindow");
}
```
As you can see no window.Show() is **not** neccessary as the GetWindow already facilitates that.
Maybe it is enough to change the OpenWindow in the lib to not use Show(), but as this is part of the compiled lib I could not test it.


TLDR:
To HotFix apply the following changes to FastScriptReloadWelcomeScreen.cs
```
         public static FastScriptReloadWelcomeScreen Init()
         {
+            EditorWindow window = GetWindow(typeof(FastScriptReloadWelcomeScreen),true, _WindowTitle, true);
+            window.minSize = _WindowSizePx;
+            window.maxSize = _WindowSizePx;
+            return window as FastScriptReloadWelcomeScreen;
-             return OpenWindow<FastScriptReloadWelcomeScreen>(_WindowTitle, _WindowSizePx);
         }
```
and
```csharp
         public static void SetCommonMaterialsShader(ShadersMode newShaderModeValue)
         {
+            return;
             var rootToolFolder = AssetPathResolver.GetAssetFolderPathRelativeToScript(ScriptableObject.CreateInstance(typeof(FastScriptReloadWelcomeScreen)), 1);
             if (rootToolFolder.Contains("/Scripts"))
             {
```

**Comment by handzlikchris** [comment]
Thanks - that's great investigation. I'll look at changing the underlying code in util lib

**Comment by RunninglVlan** [comment]
I checked and found that `FastScriptReloadWelcomeScreen` is also duplicated on each recompile. So this is memory leak issue too. Hopefully it will be fixed soon!
As I tested, this also fixes issue:
```diff
public static void SetCommonMaterialsShader(ShadersMode newShaderModeValue)
{
-   var rootToolFolder = AssetPathResolver.GetAssetFolderPathRelativeToScript(ScriptableObject.CreateInstance(typeof(FastScriptReloadWelcomeScreen)), 1);
+   var welcomeScreen = ScriptableObject.CreateInstance<FastScriptReloadWelcomeScreen>();
+   var rootToolFolder = AssetPathResolver.GetAssetFolderPathRelativeToScript(welcomeScreen, 1);
+   Object.DestroyImmediate(welcomeScreen);
    if (rootToolFolder.Contains("/Scripts"))
```
Restart Editor after this change (it won't remove already added EditorWindows from memory).

**Comment by Kadagaden** [comment]
Any update on this?

**Comment by Infinite-3D** [comment]
> Any update on this?

This only started happening when I updated the package to the latest version, so I reverted back to FSR v1.6.1 and that fixed it.

**Comment by handzlikchris** [comment]
thanks - that's interesting, should be easier to narrow down. Thought it'll have to do with Unity update

**Comment by RunninglVlan** [comment]
https://github.com/handzlikchris/FastScriptReload/issues/141#issuecomment-2678225460
The issue is that temporary Scriptable object is created in memory and never destroyed.
The fix is literally 2 lines of code, unless someone reviews it and says that it's not the case.
I think it would be more productive to fix it than adding more comments.
